scalable-go-service/index.md

```diff
@@ -0,0 +1,322 @@
+
+Two of our main hotel vendors have announced major changes to the APIs we use to
+load hotel rates. We saw this as an opportunity to reevaluate and improve the
+backend services that interact with these APIs. To do this, we need to use
+different strategies to load rates based on the vendor and their API contracts.
+Most notably, some vendors allow us to get all available rates for multiple
+hotels in a single request while others only give data on an individual hotel
+basis.
+
+The latter case causes a lot of bottleneck for the shopping flow because even
+though the individual requests resolve fairly quickly, the time complexity grows
+linearly with the number of hotels. This would scale poorly as we increase the
+number of hotels we search. It also hinders user experience and confidence in
+the service if the user has to wait several seconds or even minutes for the
+search results to come back.
+
+## Concurrency and Go
+
+Since we have no influence on how our vendors' APIs are designed, it was clear
+to the team that we had to parallelize our rates requests for individual hotels.
+We were already parallelizing these requests in our services that interact with
+the sunsetting APIs. The original implementation used Javascript promises to
+create multiple concurrent calls for hotel rates. We wanted to evaluate this
+implementation and look for ways to improve it by using a more suitable
+programming language or paradigm that could help the efficiency and design of
+our concurrent program.
+
+If you are not familiar, `go` is a relatively new language and is designed based
+on modern understanding of parallel computing. Its primitive types make
+designing and building concurrent applications simple and fun. A `goroutine` can
+be thought of an inexpensive and lightweight thread. A `channel` enables two or
+more routines to communicate with each other, similar to a pipe. The language
+also provides a multiway concurrent control switch through the `select`
+statement, which enables interactions with multiple `channels`. So, creating a
+`goroutine` is cheap and we are able to use first-class primitives to facilitate
+conversation between them. But how do we use these features to design and
+implement a scalable service for gathering hotel rates?
+
+## Communicating Sequential Processes
+
+Concurrency in `go` is based on Tony Hoare's CSP, which specifies the patterns
+of interactions between concurrent processes.  Unlike the traditional thread
+model, CSP takes a message passing approach to ensure data integrity rather than
+locks and mutexes. Go's `goroutines` and `channels` mediate access to shared
+data by passing data references between processes and help structure concurrent
+software. Independent components should run concurrently and communicate with
+other components, if need be. This decouples concurrency from parallelism and
+makes programs easier to digest. Concurrency is the composition of independently
```

Comment:
Does explicitly distinguishing between concurrency and parallelism help make your overall point / tell a better story?